py3status/modules/imap_idle.py

```diff
@@ -0,0 +1,152 @@
+# -*- coding: utf-8 -*-
+"""
+Display number of unread messages from IMAP account. (using IMAP4 IDLE)
+
+Configuration parameters:
+    allow_urgent: display urgency on unread messages (default False)
+    criterion: status of emails to check for (default 'UNSEEN')
+    format: display format for this module (default 'Mail: {unseen}')
+    hide_if_zero: hide this module when no new mail (default False)
+    mailbox: name of the mailbox to check (default 'INBOX')
+    password: login password (default None)
+    port: number to use (default '993')
+    security: login authentication method: 'ssl' or 'starttls'
+        (startssl needs python 3.2 or later) (default 'ssl')
+    server: server to connect (default None)
+    user: login user (default None)
+
+Format placeholders:
+    {unseen} number of unread emails
+
+Color options:
+    color_new_mail: use color when new mail arrives, default to color_good
+
+@author obb and girst
+
+SAMPLE OUTPUT
+{'full_text': 'Mail: 36', 'color': '#00FF00'}
+"""
+from imaplib2 import imaplib2
+from threading import Thread
+from ssl import create_default_context
+STRING_UNAVAILABLE = 'N/A'
+
+
+class Py3status:
+    """
+    """
+    # available configuration parameters
+    allow_urgent = False
+    criterion = 'UNSEEN'
+    format = 'Mail: {unseen}'
+    hide_if_zero = False
+    mailbox = 'INBOX'
+    password = None
+    port = '993'
+    security = 'ssl'
+    server = None
+    user = None
+
+    class Meta:
+        deprecated = {
+            'rename': [
+                {
+                    'param': 'new_mail_color',
+                    'new': 'color_new_mail',
+                    'msg': 'obsolete parameter use `color_new_mail`',
+                },
+                {
+                    'param': 'imap_server',
+                    'new': 'server',
+                    'msg': 'obsolete parameter use `server`',
+                },
+            ],
+        }
+
+    def post_config_hook(self):
+        self.mail_count = 0 # will be used to store
+        if self.security not in ["ssl", "starttls"]:
+            raise ValueError("Unknown security protocol")
+        self.idle_thread = self.Idle_Thread(
+            mailbox=self.mailbox,
+            criterion=self.criterion,
+            user=self.user,
+            password=self.password,
+            server=self.server,
+            port=self.port,
+            security=self.security)
+        self.idle_thread.start()
+
+    class Idle_Thread(Thread):
```

Comment:
I'm not familiar with threads. I'm wondering if we need `class` here. If you look at  `scratchpad_async`, there is a `_listen` method that might be doing the similar thing here without the class. Idk. Maybe just play with it.

Reply:
that's how i've always done them, but indeed, not necessary here (actually, encapsulation gets annoying here)

changed. also merged synchronous and async version of `_get_mail_count`